Refactor CMake system: simplify target definitions and other configuration

halleysfifthinc opened the pull request:
Ok this is the major build system refactor! Net -1631/+653 lines (removed/added) lines of CMake configuration code, and 7 files deleted. Externally visible changes (e.g. during configuration or to the build products) are minimal and listed below.

In #789, it was suggested that the potentially breaking changes to the build options should be announced for comment on the SimTK forums. I'm happy to do that once the general direction of this PR has been review and approved.

## Summary of changes:

- (potentially BREAKING) Only one set of libraries is built for a given configuration/build directory
    - Removed support for simultaneous static and shared library builds
    - Removed support for simultaneous versioned and unversioned library builds
- User configurable build options have all been moved to the top of `CMakeLists.txt` and available options have been reduced to improve clarity and eliminate redundancy
- Target properties (link options, compile definitions, include directories, etc) have been applied directly to targets instead of at the project/directory level
- Simplified installation of vendored/runtime artifact (e.g. BLAS/LAPACK shared libraries, glut)
- General cleanup of dead CMake code

### Changes to the build system:

#### DEPRECATED build variables:
- `BUILD_DYNAMIC_LIBRARIES` and `BUILD_STATIC_LIBRARIES`
  - Replaced with new variable `SIMBODY_BUILD_SHARED_LIBS` (or `BUILD_SHARED_LIBS`) (default: true) to control library type (static or dynamic/shared)
- `BUILD_UNVERSIONED_LIBRARIES`
  - Replaced with (already existing) variable `BUILD_VERSIONED_LIBRARIES` (default: false) to control whether library file names include the version.
- `BUILD_TESTS_AND_EXAMPLES_STATIC` and `BUILD_TESTS_AND_EXAMPLES_SHARED`
  - Tests and/or examples are built against whichever type of library is set by `SIMBODY_BUILD_SHARED_LIBS`

Use of deprecated build variables will trigger a warning and then the requested setting will be applied to the relevant variable if appropriate. Conflicting build option settings (e.g. `-DBUILD_DYNAMIC_LIBRARIES=ON -DBUILD_STATIC_LIBRARIES=ON`) will trigger a fatal error.

#### New build variables:

- (advanced) `SIMBODY_STATIC_LIBRARIES_POSTFIX::String` Allows replicating previous (unusual) behaviour to add a suffix (e.g. `"_static"`) for static library builds 

### Changes to the build artifacts (install zip's):

- Platform agnostic changes:
    - Renamed library target names (from unifying static/shared builds) required changes to `SimbodyConfig.cmake`
      - Target names are now hard-coded to the previous default names (e.g. `SimTKcommon`, `SimTKmath`, etc) so this change is functionally invisible for typical builds
    - Things that were originally intended to be set as public/interface linking options changed `SimbodyTargets.cmake`
      - Public/interface definition `SimTK_USE_STATIC_LIBRARIES`
      - Public/interface option requiring C++11 language standard
      - Compile option for MSVC compiler to silence Debug warning (`/wd4996`)
- Changes to the Windows install zip:
  - Removal of unused `dirent.h`

## Testing I have performed:

- Built and passed tests on Linux and Windows
- Built OpenSim and passed tests on Linux and Windows

I have also reviewed the conda-forge, vcpkg, and several distribution build recipes:
- conda-forge:
  - simbody will no longer install dirent.h, which removes the need to [patch simbody](https://github.com/conda-forge/simbody-feedstock/blob/716dfa65bb306e308ae5408ea4184cc31347a3c6/recipe/716.patch) and [install dirent as a dependency](https://github.com/conda-forge/simbody-feedstock/blob/716dfa65bb306e308ae5408ea4184cc31347a3c6/recipe/meta.yaml#L78C9-L78C26)
- vcpkg:
  - Consistent static/shared library names removes the need to [patch simbody](https://github.com/microsoft/vcpkg/blob/master/ports/simbody/common-name-libs.patch)
  - Simultaneous static/shared builds were not supported, but use of `BUILD_DYNAMIC_LIBRARIES` and `BUILD_STATIC_LIBRARIES` will need to be updated
- Debian based: unaffected
- Arch Linux AUR: unaffected
- FreeBSD: unaffected

Remaining todo:
- Update build instructions in README (waiting for the general direction/contents of this PR to be approved and then I will add it here or in a separate PR, depending on preferences).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/simbody/simbody/801)
<!-- Reviewable:end -->

aymanhab commented:
Thanks for checking @sherm1 and indeed we only use the options described by @nickbianco.  It'd also be good to build on osx as it has its own quirks, also not sure what options are used by the conda recipe of simbody or gazebo/other clients of simbody. 
From our perspective if it builds successfully and we can build OpenSim with the artifact on all platforms  (with/without visualizer) then we're good.

halleysfifthinc commented:
FYI the latest Gazebo version doesn't depend on Simbody. Only "Gazebo-classic" depends on Simbody; Gazebo-classic is still being maintained but not actively developed, and the README directs users to the new version.

I discussed changes needed to downstream clients/recipes at the bottom of the initial PR message.

aymanhab commented:
Thanks @halleysfifthinc for the thorough job chasing the various potential clients, it seems we're covered then 👍

sherm1 reviewed (commented):
Thanks, Allen! I didn't have time yet for a full review but had a question.

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/801#-:-O8bieqQ3Wykz8Lo90GS:bjh7sjz)* status: 0 of 29 files reviewed, 2 unresolved discussions (waiting on @halleysfifthinc)

___
*[`CMakeLists.txt` line 37 at r1](https://reviewable.io/reviews/simbody/simbody/801#-O8bi5S2-DwCWtLEYVeb:-O8bi5S3-egV6E_0Eku7:bs5xd95) ([raw file](https://github.com/simbody/simbody/blob/9e967a3630af3f862cca9aabf4516f9cd4cc1c81/CMakeLists.txt#L37)):*
> ```CMake
>     set(shared_lib_default ${BUILD_SHARED_LIBS})
> endif()
> option(SIMBODY_BUILD_SHARED_LIBS
> ```

BTW since the default is to build shared libraries (and I think most users prefer those), did you consider having the variable be `SIMBODY_BUILD_STATIC_LIBS` instead? That's the odd case so maybe it would make more sense to have a variable for positively choosing that rather than disabling shared libs?
___
*[`CMakeLists.txt` line 332 at r1](https://reviewable.io/reviews/simbody/simbody/801#-O8bhcDfCGCgMyAOsh0s:-O8bhcDfCGCgMyAOsh0t:b830qem) ([raw file](https://github.com/simbody/simbody/blob/9e967a3630af3f862cca9aabf4516f9cd4cc1c81/CMakeLists.txt#L332)):*
> ```CMake
> else()
>     set(default_build_inst_set)
> endif()
> ```

nit: the above if/else seems unnecessary


<!-- Sent from Reviewable.io -->

halleysfifthinc commented:
(Responding here since logging into the Reviewable website doesn't work for me.)

> did you consider having the variable be `SIMBODY_BUILD_STATIC_LIBS` instead?

I used `SIMBODY_BUILD_SHARED_LIBS` as the project level counterpart to the standard/global CMake variable [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html). I agree that the user experience of "[manually] enable this unusual config" can be slightly more understandable than "disable this normal config", but I think it is more important here to match modern CMake conventions and functionality as much as possible.

> nit: the above if/else seems unnecessary

Agreed. I pushed a commit that slightly simplifies that. In general, the CMake compiler flags code needs some TLC, but that was separate enough to the goals of this PR that I will wait and address that in the future.

sherm1 reviewed (commented):
My first attempt to build with this PR didn't go well -- please advise! Here's what happened:
- Running on Ubuntu 22.04
- I used CMake 3.22.1, deleted the cache, configured (took all defaults), generated.
- then in the binary directory I did `make -j40`.
- Got: `Error copying file "-t" to "/home/sherm/Documents/GitHub/simbody_binaries".`

![image.png](https://files.reviewableusercontent.io/images/reviewable/4088016/OX2m6m6Pmo90nUZdwssm27OY/image.png)

Output:
```
sherm@PUGET-180653:~/Documents/GitHub/simbody_binaries $ make -j40
-- Trying to compile with system BLAS/LAPACK libraries
-- Trying to compile with system BLAS/LAPACK libraries - success.
-- Using BLAS/LAPACK libraries BLAS::BLAS;LAPACK::LAPACK
-- Checking for module 'glut'
--   No package 'glut' found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/sherm/Documents/GitHub/simbody_binaries
[  0%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/src/About.cpp.o
[  0%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/src/Parallel2DExecutor.cpp.o
[  0%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/src/ParallelExecutor.cpp.o
[  0%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/src/NiceTypeName.cpp.o
[  0%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/src/PrivateInstantiations.cpp.o
...
[  5%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/Polynomial/src/rpoly.cpp.o
[  5%] Building CXX object SimTKcommon/CMakeFiles/SimTKcommon.dir/Simulation/src/System.cpp.o
[  6%] Linking CXX shared library ../libSimTKcommon.so
Error copying file "-t" to "/home/sherm/Documents/GitHub/simbody_binaries".
make[2]: *** [SimTKcommon/CMakeFiles/SimTKcommon.dir/build.make:707: libSimTKcommon.so.3.8] Error 1
make[1]: *** [CMakeFiles/Makefile2:1730: SimTKcommon/CMakeFiles/SimTKcommon.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/801#-:-O914_0bDCHa2MA3dm-6:b4jy8xr)* status: 0 of 29 files reviewed, 2 unresolved discussions (waiting on @halleysfifthinc)



<!-- Sent from Reviewable.io -->

halleysfifthinc commented:
Should be fixed now.

sherm1 reviewed (commented):
Thanks, Allen

The good news is that Simbody builds fine now and all the tests pass. The bad news is that `make test` took 995s to run -- previously was around 70s, a >14X slowdown! The problem appears to be that `CMAKE_BUILD_TYPE` is now blank and apparently defaulting to `Debug`. Previously the default was `RelWithDebInfo`. If I set it manually to `RelWithDebInfo` the tests run in 72s on my machine.

We don't want users first experience running Simbody to be in Debug! It's very painful that way.

> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/801#-:-O9b2_s37XHLTeszvIge:baisjot)* status: 0 of 29 files reviewed, 2 unresolved discussions (waiting on @halleysfifthinc)



<!-- Sent from Reviewable.io -->

sherm1 reviewed (approved):
`make install` also looks good. So I think this is ready to merge pending a change to the default Configuration type.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 26 of 29 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/801#-:-O9b92v735aXqCXTOQiK:bxh7g7u)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @halleysfifthinc)



<!-- Sent from Reviewable.io -->

halleysfifthinc commented:
:+1:  I won't have time to move on this until after my thesis defense near the end of this month.

sherm1 commented:
See PR #812 for the fix to set CMAKE_BUILD_TYPE (missing "FORCE"). That should be included here also when you have a chance. I hope your defense went well, @halleysfifthinc !

halleysfifthinc commented:
Thanks @sherm1, my defense was successful! I've fixed the CMAKE_BUILD_TYPE default and updated the build instructions in the README. This should now be ready to merge!

A while back, you had suggested announcing the changes (removal, deprecation) of some of the build variables on the SimTK forums. Is that something you're still wanting done before merging?

sherm1 commented:
Congratulations, Dr Hill!

No need to announce before merging. I'll review, then merge, then you can post something there.

sherm1 reviewed (approved):
I built and installed on Ubuntu 22.04 and verified that the default is optimized. All tests pass and installation looks correct. Thanks @halleysfifthinc! I'll squash and merge into master now. If you want to post a note on the Simbody forum so people will know what to look for, please do.

> Reviewed 9 of 9 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/simbody/simbody/801#-:-OKOSNs71iaCyVh5nbqx:bhjvwtd)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @halleysfifthinc)



<!-- Sent from Reviewable.io -->
